language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -20,44 +24,63 @@ import           Language.PlutusCore.Lexer.Type hiding (name)
 import           Language.PlutusCore.Name
 import           Language.PlutusCore.Normalize
 import           Language.PlutusCore.Quote
-import           Language.PlutusCore.Renamer    (annotateType, rename)
+import           Language.PlutusCore.Renamer    (rename)
 import           Language.PlutusCore.Type
 import           PlutusPrelude
 
+import           Control.Lens                   (coerced)
+import           Control.Lens.TH                (makeLenses)
 import           Control.Monad.Error.Lens
 import           Control.Monad.Except
 import           Control.Monad.Reader
+import qualified Data.IntMap                    as IM
 import           Data.Map                       (Map)
 import qualified Data.Map                       as Map
 
 {- Note [Costs]
-Typechecking costs are relatively simple: it costs 1 gas to perform
-a reduction. Substitution does not in general cost anything.
+Typechecking costs are relatively simple: it costs 1 gas to perform a reduction.
+Substitution does not in general cost anything.
 
 Costs are reset every time we enter 'NormalizeTypeT'.
 
-In unlimited mode, gas is not tracked and we do not fail even on large numbers
-of reductions.
+In unlimited mode, gas is not tracked and we do not fail even on large numbers of reductions.
 -}
 
 -- | Mapping from 'DynamicBuiltinName's to their 'Type's.
 newtype DynamicBuiltinNameTypes = DynamicBuiltinNameTypes
-    { unDynamicBuiltinNameTypes :: Map DynamicBuiltinName (Quote (Type TyName ()))
+    { unDynamicBuiltinNameTypes :: Map DynamicBuiltinName (NormalizedType TyName ())
     } deriving (Semigroup, Monoid)
 
+-- TODO: define @UniqueMap@ already.
```

Comment:
:+1:

Reply:
I'll do this in a subsequent PR as it touches a lot of other parts of code.